cpp/src/io/comp/uncomp.cpp

```diff
@@ -726,30 +638,55 @@ void host_decompress(compression_type compression,
   }
 }
 
-[[nodiscard]] bool use_host_decompression(compression_type compression, size_t num_buffers)
+enum class host_engine_state : uint8_t { ON, OFF, AUTO };
+
+[[nodiscard]] host_engine_state get_host_engine_state(compression_type compression)
 {
   auto const has_host_support   = is_host_decompression_supported(compression);
   auto const has_device_support = is_device_decompression_supported(compression);
   CUDF_EXPECTS(has_host_support or has_device_support,
                "Unsupported compression type: " + compression_type_name(compression));
-  if (not has_host_support) { return false; }
-  if (not has_device_support) { return true; }
+  if (not has_host_support) { return host_engine_state::OFF; }
+  if (not has_device_support) { return host_engine_state::ON; }
 
   // If both host and device compression are supported, dispatch based on the environment variable
   auto const env_var = getenv_or("LIBCUDF_HOST_DECOMPRESSION", std::string{"OFF"});
+
   if (env_var == "AUTO") {
-    auto const threshold =
-      getenv_or("LIBCUDF_HOST_DECOMPRESSION_THRESHOLD", default_host_decompression_auto_threshold);
-    return num_buffers < threshold;
+    return host_engine_state::AUTO;
+  } else if (env_var == "OFF") {
+    return host_engine_state::OFF;
+  } else if (env_var == "ON") {
+    return host_engine_state::ON;
+  }
+  CUDF_FAIL("Invalid LIBCUDF_HOST_DECOMPRESSION value: " + env_var);
```

Comment:
I think `CUDF_UNREACHABLE` is better, since it can suppress compiler warning about ending function without return value.